script/CheckIsAtom.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE ScopedTypeVariables, StandaloneDeriving, DeriveFunctor #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+-- For a description of this program see http://neilmitchell.blogspot.co.uk/2018/02/atomic-expressions-generically.html
+module Main(main) where
+
+-- Note that 1 vs -1  is one of the few things that matters!!!
+
+import Language.Haskell.Exts
+import Language.Haskell.Exts.Util
+import Control.Monad
+import Data.Data
+import System.Random
+import System.IO
+import Control.Exception
+
+
+main :: IO ()
+main = do
+    hSetBuffering stdout NoBuffering
+    let modes = defaultParseMode{extensions = map EnableExtension [minBound .. maxBound]}
+    checkAtomicity (fmap void . parseExpWithMode modes) $ \x -> App () x x
+    checkAtomicity (fmap void . parseTypeWithMode modes) $ \x -> TyApp () x x
+    checkAtomicity (fmap void . parsePatWithMode modes) $ \x -> PApp () (UnQual () $ Ident () "Foo") [x,x]
+
+
+deriving instance Eq a => Eq (ParseResult a)
+
+checkAtomicity :: forall a . (Pretty a, Eq a, Show a, Data a) => Brackets a => (String -> ParseResult a) -> (a -> a) -> IO ()
+checkAtomicity parse wrap =
+    forM_ (dataTypeConstrs $ dataTypeOf (undefined :: a)) $ \ctor -> do
+        putStr $ show (typeOf (undefined :: a)) ++ " " ++ show ctor ++ " ... "
+        ans <- replicateM 10000 $ do
+            handle (\LimitReached -> return False) $ do
+                x :: a <- fromConstrM (mkValue 50) ctor
+                if parse (prettyPrint x) /= ParseOk x then
+                    return False
+                else do
+                    let seemsAtomic = parse (prettyPrint $ wrap x) == ParseOk (wrap x)
+                    let saysAtomic = isAtom x
+                    when (not seemsAtomic && isAtom x) $
+                        putStrLn $ unlines $
+                            [""
+                            ,"DIAGREE!:"
+                            ,"  " ++ prettyPrint x
+                            ,"  " ++ show x
+                            ,"  isAtom = " ++ show saysAtomic]
+                    return $ seemsAtomic == saysAtomic
```

Comment:
Maybe this Bool needs to be strict to avoid the space leak mentioned in the blog post ?

Reply:
There is a space leak of 10K bools here, which is not very much. The space leak I see is ~10Gb, so it will be a different one.